File: autozig-math/zig/float_ord.zig
const std = @import("std");

/// A wrapper around `f32` that provides ordering and hashing
pub const FloatOrd = extern struct {
    value: f32,

    pub fn new(value: f32) FloatOrd {
        return .{ .value = value };
    }

    /// Compare for ordering (handles NaN by treating it as greater than all other values)
    pub fn cmp(self: FloatOrd, other: FloatOrd) i32 {
        const a = self.value;
        const b = other.value;

        // Handle NaN cases
        const a_nan = std.math.isNan(a);
        const b_nan = std.math.isNan(b);

        if (a_nan and b_nan) return 0;
        if (a_nan) return 1;
        if (b_nan) return -1;

        // Normal comparison
        if (a < b) return -1;
        if (a > b) return 1;

        // Handle -0.0 vs 0.0
        const a_bits = @as(u32, @bitCast(a));
        const b_bits = @as(u32, @bitCast(b));
        if (a_bits < b_bits) return -1;
        if (a_bits > b_bits) return 1;
        return 0;
    }

    pub fn eq(self: FloatOrd, other: FloatOrd) bool {
        return self.cmp(other) == 0;
    }

    /// Hash the float value
    pub fn hash(self: FloatOrd) u32 {
        return @as(u32, @bitCast(self.value));
    }
};

export fn float_ord_new(value: f32) FloatOrd {
    return FloatOrd.new(value);
}

export fn float_ord_cmp(self: FloatOrd, other: FloatOrd) i32 {
    return self.cmp(other);
}

export fn float_ord_hash(self: FloatOrd) u32 {
    return self.hash();
}

File: autozig-math/src/float_ord.rs
use autozig::include_zig;
use std::cmp::Ordering;
use std::hash::{Hash, Hasher};

#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub struct FloatOrd(pub f32);

include_zig!("zig/float_ord.zig", {
    fn float_ord_new(value: f32) -> FloatOrd;
    fn float_ord_cmp(self_: FloatOrd, other: FloatOrd) -> i32;
    fn float_ord_hash(self_: FloatOrd) -> u32;
});

impl FloatOrd {
    pub fn new(value: f32) -> Self {
        float_ord_new(value)
    }
}

impl PartialEq for FloatOrd {
    fn eq(&self, other: &Self) -> bool {
        float_ord_cmp(*self, *other) == 0
    }
}

impl Eq for FloatOrd {}

impl PartialOrd for FloatOrd {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for FloatOrd {
    fn cmp(&self, other: &Self) -> Ordering {
        match float_ord_cmp(*self, *other) {
            -1 => Ordering::Less,
            0 => Ordering::Equal,
            _ => Ordering::Greater,
        }
    }
}

impl Hash for FloatOrd {
    fn hash<H: Hasher>(&self, state: &mut H) {
        float_ord_hash(*self).hash(state);
    }
}

impl From<f32> for FloatOrd {
    fn from(value: f32) -> Self {
        FloatOrd::new(value)
    }
}
